### ISSUE TITLE
Validation rules for Client Controlled Nullability operators

### ISSUE THREAD

**Issue by aprilrd** [aprilrd/graphql-js#6]
Here is a prroposed list of validations with nullability operators. I am sure there are cases that I missed here so please comment below if you see any cases missed. The current changeset on graphql-js was made with this list in mind, but we need to audit and make sure they behave per this list.

# Schema + Operation validation
All the tests under this section assumes this following schema
```graphql
interface SomeBox {
  deepBox: SomeBox
  unrelatedField: String
}

type StringBox implements SomeBox {
  scalar: String
  deepBox: StringBox
  unrelatedField: String
  listStringBox: [StringBox]
  stringBox: StringBox
  intBox: IntBox
}

type IntBox implements SomeBox {
  scalar: Int
  deepBox: IntBox
  unrelatedField: String
  listStringBox: [StringBox]
  stringBox: StringBox
  intBox: IntBox
}

interface NonNullStringBox1 {
  scalar: String!
}

type NonNullStringBox1Impl implements SomeBox & NonNullStringBox1 {
  scalar: String!
  unrelatedField: String
  deepBox: SomeBox
}

interface NonNullStringBox2 {
  scalar: String!
}

type NonNullStringBox2Impl implements SomeBox & NonNullStringBox2 {
  scalar: String!
  unrelatedField: String
  deepBox: SomeBox
}

type Connection {
  edges: [Edge]
}

type Edge {
  node: Node
}

type Node {
  id: ID
  name: String
}

type Query {
  someBox: SomeBox
  connection: Connection
}
```

#### With aliases
- [ ] _It's unclear to me what the behavior should be. Probably invalid?_

##### Current Status

- There is a test

**Current test case**
```graphql
"""
currently asserting validity
"""
{
  someBox {
    ...on SomeBox {
      nullable: unrelatedField
    }
    ...on IntBox {
      nonNullable: unrelatedField!
    }
  }
}
```

#### Without aliases
&nbsp; | ! | ? | not-marked
-- | -- | -- | --
! | valid (case 1) | (duplicate) | (duplicate)
? | invalid (case 2) | valid (case 3) | (duplicate)
not-marked | invalid (case 4) | valid (case 5) | valid (case 6)

##### Current Status

- Case 1 is implemented
- Case 4 is implemented

**Sample test cases**
```graphql
"""
case 1: valid
"""
{
  someBox {
    ...on SomeBox {
      unrelatedField!
    }
    ...on IntBox {
      unrelatedField!
    }
  }
}

"""
case 4: invalid
"""
{
  someBox {
    ...on SomeBox {
      unrelatedField
    }
    ...on IntBox {
      unrelatedField!
    }
  }
}
```

#### With aliases

- [ ] _It's unclear to me what the behavior should be. Probably invalid?_

##### Current Status

- There is a test

**Current Test Case**

```graphql
"""
currently asserting validity
"""
{
  someBox {
    ...on NonNullStringBox1 {
      nonNullable: scalar!
    }
    ...on NonNullStringBox1Impl {
      nullable: scalar
    }
  }
}
```

#### Both with or without aliases

&nbsp; | ! | ? | not-marked
-- | -- | -- | --
! | valid (case 1) | (duplicate) | (duplicate)
? | invalid (case 2) | valid (case 3) | (duplicate)
not-marked | valid (case 4) | valid (case 5) | valid (case 6)

##### Current Status

- Case 1 is implemented
- Case 4 is implemented

**Sample test cases**
```graphql
"""
case 1: valid
"""
{
  someBox {
    ...on NonNullStringBox1 {
      scalar!
    }
    ...on NonNullStringBox1Impl {
      scalar!
    }
  }
}

"""
case 4: valid
"""
{
  someBox {
    ...on NonNullStringBox1 {
      scalar
    }
    ...on NonNullStringBox1Impl {
      scalar!
    }
  }
}
```

## When types are exclusive
Here, with aliases, all combinations are valid. 

- [ ] Do a double pass to ensure that marking a nullable field non-nullable stays backward compatible.

### Fields are both nullable
&nbsp; | ! | ? | not-marked
-- | -- | -- | --
! | valid (case 1) | (duplicate) | (duplicate)
? | invalid (case 2) | valid (case 3) | (duplicate)
not-marked | invalid (case 4) | valid (case 5) | valid (case 6)

#### Current Status

- Case 4 is implemented

**Sample test cases**
```graphql
"""
case 1: valid
"""
{
  someBox {
    ... on IntBox {
      scalar!
    }
    ... on StringBox {
      scalar!
    }
  }
}

"""
case 4: invalid
"""
{
  someBox {
    ... on IntBox {
      scalar
    }
    ... on StringBox {
      scalar!
    }
  }
}
```

### Fields are both non-nullable

&nbsp; | ! | ? | not-marked
-- | -- | -- | --
! | valid | (duplicate) | (duplicate)
? | invalid | valid | (duplicate)
not-marked | valid | invalid | valid

#### Current Status

- Not implemented

**Sample test cases**
The current schema does not have non-nullable fields across exclusive types. We need to add that first.

### Field A is nullable and Field B is non-nullable
&nbsp; | ! on B | ? on B | Not-marked on B
-- | -- | -- | --
! on A | valid (case 1) | invalid (case 2) | valid (case 3)
? on A | invalid (case 4) | valid (case 5) | invalid (case 6)
Not-marked A | invalid (case 7) | valid (case 8) | invalid (case 9)

#### Current Status

- Case 3 is implemented

**Sample test cases**
```graphql
"""
case 1: valid
"""
{
  someBox {
    ... on StringBox {
      scalar!
    }
    ... on NonNullStringBox1 {
      scalar!
    }
  }
}

"""
case 3: valid
"""
{
  someBox {
    ... on StringBox {
      scalar!
    }
    ... on NonNullStringBox1 {
      scalar
    }
  }
}

"""
case 8: valid
"""
{
  someBox {
    ... on StringBox {
      scalar
    }
    ... on NonNullStringBox1 {
      scalar?
    }
  }
}

```

# Operation only validation
Given fragment A and fragment B (one of them could just be a selection set)

&nbsp; | ! on B | ? on B | Not-marked on B
-- | -- | -- | --
! on A | valid (case 1) | (duplicate) | (duplicate)
? on A | invalid (case 2) | valid (case 3) | (duplicate)
Not-marked A | valid (case 4) | valid (case 5) | valid (case 6)

## Current Status

- [ ] _We have a test for Case 4, but it asserts `invalid`. I believe it should be valid. Open for feedback_

**Sample test cases**
```graphql
"""
case 1: valid
"""
fragment conflictingNullability on Pet {
  ... on Dog {
    name!
  }
  ... on Cat {
    name!
  }
}

"""
case 2: invalid
"""
fragment conflictingNullability on Pet {
  ... on Dog {
    name?
  }
  ... on Cat {
    name!
  }
}
```

**Comment by magicmark** [comment]
🎉 thanks for this! this is super helpful in working out the corner cases :)

I'm having a little trouble understand the tables in context - maybe we could add example schema + queries / fragments and label the axes?

(I took a stab at it but then deleted my comments cos i realized i was being dumb :P)

**Comment by magicmark** [comment]
> ### When a field is nullable
> #### With, or without aliases
> &nbsp; | ! | ? | not-marked
> -- | -- | -- | --
> ! | valid | (duplicate) | (duplicate)
> ? | invalid | valid | (duplicate)
> not-marked | valid | valid | valid

ok here's my stab - maybe something like:

### When a field is nullable

**Schema**

```graphql
type Query {
  greeting: String
}
```

**Query**

```graphql
{
  greeting
}
```

#### With, or without aliases

&nbsp; | `Query.greeting!` (schema) | `Query.greeting?` (schema) | `Query.greeeting` not-marked (schema)
-- | -- | -- | --
`greeting!` (query) | valid | (duplicate) | (duplicate)
`greeting?` (query) | invalid | valid | (duplicate)
`greeting` not-marked (query) | valid | valid | valid

---

Ok this is pretty gross, maybe there's a better way to actually format this - but I think some sort of visualization aid might be helpful here?

Thanks again, I think these tables to enumerate all possibilities arre super helpful to work out the subtleties

**Comment by magicmark** [comment]
Actually I'm being dumb aren't I - the two axes are both for fragments? (What even is `Query.greeting?`, that's not a thing in SDL or being proposed right?)

**Comment by aprilrd** [comment]
Definitely, I will add more labels to the tables. I also liked your suggestion to add some examples to ground the cases in the tables so I will do that as well.

**Comment by aprilrd** [comment]
Updated the issue with examples and added the current status of tests we have. I marked open questions with the checkboxes.
